client/electron/go_vpn_tunnel.ts

```diff
@@ -246,27 +255,44 @@ class GoTun2socks {
       args.push('-dnsFallback');
     }
 
-    this.stopRequested = false;
-    let autoRestart = false;
-    do {
-      if (autoRestart) {
-        console.warn(`tun2socks exited unexpectedly. Restarting...`);
-      }
-      autoRestart = false;
-      this.process.onStdErr = (data?: string | Buffer) => {
+    const whenProcessEnded = this.launchWithAutoRestart(args);
+
+    // Either started successfully, or terminated exceptionally
+    return Promise.race([this.whenStarted, whenProcessEnded]);
+  }
+
+  private monitorStarted(): Promise<void> {
+    return (this.whenStarted = new Promise(resolve => {
+      this.process.onStdOut = (data?: string | Buffer) => {
         if (data?.toString().includes('tun2socks running')) {
```

Comment:
A better way would be refactoring to use rest API, with server side event enabled, so the client can receive "launched" event from the server.

Reply:
I added a comment to the corresponding Go code.

I feel a magic constant doesn't help a lot because it's only used once here and `const TUN2SOCKS_STARTED_LOG_MSG = 'tun2socks running'` doesn't really add anything.

Reply:
Ah my bad, I somehow thought that was also in TypeScript. Nevermind! Thanks for adding the comment though.